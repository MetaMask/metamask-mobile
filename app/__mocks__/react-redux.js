export * from 'react-redux-test';
