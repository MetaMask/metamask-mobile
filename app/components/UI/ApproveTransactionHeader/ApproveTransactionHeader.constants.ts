import AppConstants from '../../../core/AppConstants';

// eslint-disable-next-line import/prefer-default-export
export const MAX_DOMAIN_TITLE_LENGTH = 18;
export const { ORIGIN_DEEPLINK, ORIGIN_QR_CODE } = AppConstants.DEEPLINKS;

export const FAV_ICON_URL = (hostUrl: string) =>
  `https://api.faviconkit.com/${hostUrl}/64`;
