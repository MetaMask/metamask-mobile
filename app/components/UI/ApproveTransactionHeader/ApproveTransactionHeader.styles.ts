// Third party dependencies.
import { StyleSheet } from 'react-native';
import { Theme } from '../../../util/theme/models';
import { fontStyles } from '../../../styles/common';

/**
 * Style sheet function for ModalConfirmation component.
 *
 * @param params Style sheet params.
 * @param params.theme App theme from ThemeContext.
 * @param params.vars Inputs that the style sheet depends on.
 * @returns StyleSheet object.
 */
const styleSheet = (params: { theme: Theme }) => {
  const { theme } = params;
  const { colors } = theme;
  return StyleSheet.create({
    transactionHeader: {
      justifyContent: 'center',
      alignItems: 'center',
    },
    domainUrlContainer: {
      flexDirection: 'row',
      alignItems: 'center',
      justifyContent: 'center',
      width: 222,
      height: 44,
      borderWidth: 1,
      borderRadius: 222 / 2,
      borderColor: colors.border.default,
      marginHorizontal: 77,
      marginBottom: 24,
    },
    iconContainer: {
      marginRight: 8,
    },
    domainUrl: {
      ...fontStyles.bold,
      textAlign: 'center',
      fontSize: 14,
      color: colors.text.default,
    },
  });
};

export default styleSheet;
