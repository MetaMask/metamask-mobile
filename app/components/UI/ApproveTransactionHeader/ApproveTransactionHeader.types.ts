import { ImageSourcePropType } from 'react-native';

export interface ApproveTransactionHeaderI {
  spenderAddress: string;
  origin: string;
  url: string;
  currentEnsName?: string;
}

export interface AccountInfoI {
  balance: number;
  currency: string;
  accountName: string;
  networkName: string;
  networkImage: ImageSourcePropType | null;
}

export interface OriginsI {
  isOriginDeepLink: boolean;
  isOriginWalletConnect: boolean;
  isOriginMMSDKRemoteConn: boolean;
}
