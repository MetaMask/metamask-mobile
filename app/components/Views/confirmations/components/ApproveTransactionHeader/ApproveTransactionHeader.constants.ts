import AppConstants from '../../../../../core/AppConstants';
export const { ORIGIN_DEEPLINK, ORIGIN_QR_CODE } = AppConstants.DEEPLINKS;
export const APPROVE_TRANSACTION_ORIGIN_PILL =
  'approve_transaction_origin_pill';
