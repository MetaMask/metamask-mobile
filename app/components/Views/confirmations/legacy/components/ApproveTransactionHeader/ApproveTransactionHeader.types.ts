interface Asset {
  isETH?: boolean;
  tokenId?: string;
  address: string;
  symbol: string;
  decimals: number;
  image?: string;
  name?: string;
  standard?: string;
}

export interface ApproveTransactionHeaderI {
  chainId?: string;
  networkClientId?: string;
  origin?: string;
  url: string;
  currentEnsName?: string;
  from: string;
  asset: Asset;
  dontWatchAsset?: boolean;
  sdkDappMetadata?: {
    url: string;
    icon: string;
  };
}
