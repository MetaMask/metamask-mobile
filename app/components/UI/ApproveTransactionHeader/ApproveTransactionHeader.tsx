import React, { useState, useEffect, useMemo } from 'react';
import { View } from 'react-native';

import { getHost } from '../../../util/browser';
import { useSelector } from 'react-redux';
import networkList from '../../../util/networks';

import { renderShortAddress, renderAccountName } from '../../../util/address';
import { WALLET_CONNECT_ORIGIN } from '../../../util/walletconnect';
import { MM_SDK_REMOTE_ORIGIN } from '../../../core/SDKConnect';
import AccountBalance from '../../../component-library/components-temp/Accounts/AccountBalance';
import { renderFromWei, hexToBN } from '../../../util/number';
import { getTicker } from '../../../util/transactions';
import { TEST_REMOTE_IMAGE_SOURCE } from '../../../component-library/components-temp/Accounts/AccountBalance/AccountBalance.constants';
import Avatar, {
  AvatarVariants,
} from '../../../component-library/components/Avatars/Avatar';
import Text from '../../../component-library/components/Texts/Text';
import { BadgeVariants } from '../../../component-library/components/Badges/Badge/Badge.types';
import { strings } from '../../../../locales/i18n';
import { useStyles } from '../../../component-library/hooks';
import stylesheet from './ApproveTransactionHeader.styles';
import {
  FAV_ICON_URL,
  MAX_DOMAIN_TITLE_LENGTH,
  ORIGIN_DEEPLINK,
  ORIGIN_QR_CODE,
} from './ApproveTransactionHeader.constants';
import {
  AccountInfoI,
  ApproveTransactionHeaderI,
  OriginsI,
} from './ApproveTransactionHeader.types';

const ApproveTransactionHeader = ({
  spenderAddress,
  origin,
  url,
  currentEnsName,
}: ApproveTransactionHeaderI) => {
  const [accountInfo, setAccountInfo] = useState<AccountInfoI>({
    balance: 0,
    currency: '',
    accountName: '',
    networkName: '',
  });
  const [origins, setOrigins] = useState<OriginsI>({
    isOriginDeepLink: false,
    isOriginWalletConnect: false,
    isOriginMMSDKRemoteConn: false,
  });

  const { styles } = useStyles(stylesheet, {});

  const accounts = useSelector(
    (state: any) =>
      state.engine.backgroundState.AccountTrackerController.accounts,
  );

  const identities = useSelector(
    (state: any) =>
      state.engine.backgroundState.PreferencesController.identities,
  );

  const selectedAddress = useSelector(
    (state: any) =>
      state.engine.backgroundState.PreferencesController.selectedAddress,
  );

  const ticker = useSelector(
    (state: any) =>
      state.engine.backgroundState.NetworkController.provider.ticker,
  );

  const networkType = useSelector(
    (state: any) =>
      state.engine.backgroundState.NetworkController.provider.type,
  );

  const nickname = useSelector(
    (state: any) =>
      state.engine.backgroundState.NetworkController.provider.nickname,
  );

  useEffect(() => {
    const weiBalance = selectedAddress
      ? hexToBN(accounts[selectedAddress].balance)
      : 0;

    const balance = Number(renderFromWei(weiBalance));
    const currency = getTicker(ticker);
    const accountName = selectedAddress
      ? renderAccountName(selectedAddress, identities)
      : '';

    const isOriginDeepLink =
      origin === ORIGIN_DEEPLINK || origin === ORIGIN_QR_CODE;
    const isOriginWalletConnect = origin?.startsWith(WALLET_CONNECT_ORIGIN);

    const isOriginMMSDKRemoteConn = origin?.startsWith(MM_SDK_REMOTE_ORIGIN);

    const networkName = networkList[networkType].name;

    setAccountInfo({ balance, currency, accountName, networkName });
    setOrigins({
      isOriginDeepLink,
      isOriginWalletConnect,
      isOriginMMSDKRemoteConn,
    });
  }, [accounts, identities, origin, selectedAddress, ticker, networkType]);

  const domainTitle = useMemo(() => {
    const { isOriginDeepLink, isOriginWalletConnect, isOriginMMSDKRemoteConn } =
      origins;
    let title = '';
    if (isOriginDeepLink) title = renderShortAddress(spenderAddress);
    else if (isOriginWalletConnect)
      title = getHost(origin.split(WALLET_CONNECT_ORIGIN)[1]);
    else if (isOriginMMSDKRemoteConn) {
      title = getHost(origin.split(MM_SDK_REMOTE_ORIGIN)[1]);
    } else title = getHost(currentEnsName || url || origin);

    title =
      title.length > MAX_DOMAIN_TITLE_LENGTH
        ? `${title.substring(0, MAX_DOMAIN_TITLE_LENGTH - 3)}...`
        : title;

    return title;
  }, [currentEnsName, origin, origins, spenderAddress, url]);

  const favIconUrl = useMemo(() => {
    const { isOriginWalletConnect, isOriginMMSDKRemoteConn } = origins;
    let newUrl = url;
    if (isOriginWalletConnect) {
      newUrl = origin.split(WALLET_CONNECT_ORIGIN)[1];
    } else if (isOriginMMSDKRemoteConn) {
      newUrl = origin.split(MM_SDK_REMOTE_ORIGIN)[1];
    }
    return FAV_ICON_URL(getHost(newUrl));
  }, [origin, origins, url]);

  const renderTitle = () => (
    <View style={styles.domainUrlContainer}>
      <View style={styles.iconContainer}>
        <Avatar
          imageSource={{ uri: favIconUrl }}
          variant={AvatarVariants.Favicon}
        />
      </View>
      <Text style={styles.domainUrl}>{domainTitle}</Text>
    </View>
  );

  return (
    <View style={styles.transactionHeader}>
      {renderTitle()}

      <AccountBalance
        accountAddress={selectedAddress}
        accountNativeCurrency={accountInfo.currency}
        accountBalance={accountInfo.balance}
        accountName={accountInfo.accountName}
        accountBalanceLabel={strings('transaction.balance')}
        accountNetwork={nickname || accountInfo.networkName}
        badgeProps={{
          variant: BadgeVariants.Network,
          name: accountInfo.networkName,
          imageSource: TEST_REMOTE_IMAGE_SOURCE,
        }}
      />
    </View>
  );
};

export default ApproveTransactionHeader;
