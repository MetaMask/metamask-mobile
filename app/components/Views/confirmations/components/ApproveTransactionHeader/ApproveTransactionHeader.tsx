import { toChecksumAddress } from 'ethereumjs-util';
import React, { useEffect, useMemo, useState } from 'react';
import { View } from 'react-native';
import { useSelector } from 'react-redux';

import AppConstants from '../../../../../core/AppConstants';
import { strings } from '../../../../../../locales/i18n';
import AccountBalance from '../../../../../component-library/components-temp/Accounts/AccountBalance';
import { BadgeVariant } from '../../../../../component-library/components/Badges/Badge';
import TagUrl from '../../../../../component-library/components/Tags/TagUrl';
import { useStyles } from '../../../../../component-library/hooks';
import { selectProviderConfig } from '../../../../../selectors/networkController';
import { selectIdentities } from '../../../../../selectors/preferencesController';
import { selectAccountsByChainId } from '../../../../../selectors/accountTrackerController';
import {
  getLabelTextByAddress,
  renderAccountName,
} from '../../../../../util/address';
import { getUrlObj, prefixUrlWithProtocol } from '../../../../../util/browser';
import {
  getNetworkImageSource,
  getNetworkNameFromProviderConfig,
} from '../../../../../util/networks';
import { WALLET_CONNECT_ORIGIN } from '../../../../../util/walletconnect';
import useAddressBalance from '../../../../hooks/useAddressBalance/useAddressBalance';
import useFavicon from '../../../../hooks/useFavicon/useFavicon';
import {
  APPROVE_TRANSACTION_ORIGIN_PILL,
  ORIGIN_DEEPLINK,
  ORIGIN_QR_CODE,
} from './ApproveTransactionHeader.constants';
import stylesheet from './ApproveTransactionHeader.styles';
import { ApproveTransactionHeaderI } from './ApproveTransactionHeader.types';

const ApproveTransactionHeader = ({
  from,
  origin,
  url,
  currentEnsName,
  asset,
  dontWatchAsset,
}: ApproveTransactionHeaderI) => {
  const [accountName, setAccountName] = useState('');

  const [isOriginDeepLink, setIsOriginDeepLink] = useState(false);
  const [isOriginWalletConnect, setIsOriginWalletConnect] = useState(false);
  const [isOriginMMSDKRemoteConn, setIsOriginMMSDKRemoteConn] = useState(false);

  const { styles } = useStyles(stylesheet, {});
  const { addressBalance } = useAddressBalance(asset, from, dontWatchAsset);

  const accountsByChainId = useSelector(selectAccountsByChainId);

  const identities = useSelector(selectIdentities);
  const activeAddress = toChecksumAddress(from);

  const providerConfig = useSelector(selectProviderConfig);
  const networkName = getNetworkNameFromProviderConfig(providerConfig);

  const useBlockieIcon = useSelector(
    (state: any) => state.settings.useBlockieIcon,
  );

  useEffect(() => {
    const accountNameVal = activeAddress
      ? renderAccountName(activeAddress, identities)
      : '';

    const isOriginDeepLinkVal =
      origin === ORIGIN_DEEPLINK || origin === ORIGIN_QR_CODE;

    setAccountName(accountNameVal);
    setIsOriginDeepLink(isOriginDeepLinkVal);

    if (!origin) {
      setIsOriginWalletConnect(false);
      setIsOriginMMSDKRemoteConn(false);

      return;
    }

    setIsOriginWalletConnect(origin.startsWith(WALLET_CONNECT_ORIGIN));
    setIsOriginMMSDKRemoteConn(
      origin.startsWith(AppConstants.MM_SDK.SDK_REMOTE_ORIGIN),
    );
  }, [accountsByChainId, identities, activeAddress, origin]);

  const networkImage = getNetworkImageSource({
    networkType: providerConfig.type,
    chainId: providerConfig.chainId,
  });

  const domainTitle = useMemo(() => {
    let title = '';
    if (isOriginWalletConnect) {
      title = getUrlObj(
        (origin as string).split(WALLET_CONNECT_ORIGIN)[1],
      ).origin;
    } else if (isOriginMMSDKRemoteConn) {
      title = getUrlObj(
        (origin as string).split(AppConstants.MM_SDK.SDK_REMOTE_ORIGIN)[1],
      ).origin;
    } else {
      title = prefixUrlWithProtocol(currentEnsName || origin || url);
    }

    return title;
  }, [
    currentEnsName,
    origin,
    isOriginWalletConnect,
    isOriginMMSDKRemoteConn,
    url,
  ]);

  const faviconUpdatedOrigin = useMemo(() => {
    let newOrigin = origin as string;
    if (isOriginWalletConnect) {
      newOrigin = newOrigin.split(WALLET_CONNECT_ORIGIN)[1];
    } else if (isOriginMMSDKRemoteConn) {
      newOrigin = newOrigin.split(AppConstants.MM_SDK.SDK_REMOTE_ORIGIN)[1];
    }
    return newOrigin;
  }, [origin, isOriginWalletConnect, isOriginMMSDKRemoteConn]);

  const faviconSource = useFavicon(faviconUpdatedOrigin);

  const accountTypeLabel = getLabelTextByAddress(activeAddress);

  return (
    <View style={styles.transactionHeader}>
      {origin && !isOriginDeepLink ? (
        <TagUrl
          testID={APPROVE_TRANSACTION_ORIGIN_PILL}
          imageSource={faviconSource}
          label={domainTitle}
          style={styles.tagUrl}
        />
      ) : null}
      <AccountBalance
        accountAddress={activeAddress}
        accountTokenBalance={addressBalance}
        accountName={accountName}
        accountBalanceLabel={strings('transaction.balance')}
        accountTypeLabel={accountTypeLabel}
        accountNetwork={networkName}
        badgeProps={{
          variant: BadgeVariant.Network,
          name: networkName,
          imageSource: networkImage,
        }}
        useBlockieIcon={useBlockieIcon}
      />
    </View>
  );
};

export default ApproveTransactionHeader;
