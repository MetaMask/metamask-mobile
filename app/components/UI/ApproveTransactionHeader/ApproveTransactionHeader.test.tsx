import React from 'react';
import { Provider } from 'react-redux';
import { shallow } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import ApproveTransactionHeader from './';

const mockStore = configureMockStore();

const initialState = {
  engine: {
    backgroundState: {
      AccountTrackerController: {
        accounts: {
          '0x0': {
            balance: 200,
          },
        },
      },
      PreferencesController: {
        selectedAddress: '0x0',
        identities: {
          address: '0x0',
          name: 'Account 1',
        },
      },
      CurrencyRateController: {
        conversionRate: 10,
        currentCurrency: 'inr',
      },
      NetworkController: {
        provider: {
          ticker: 'eth',
          networkType: 'ethereum',
          nickname: 'ethereum',
        },
      },
    },
  },
};
const store = mockStore(initialState);

describe('ApproveTransactionHeader', () => {
  it('should render correctly', () => {
    const wrapper = shallow(
      <Provider store={store}>
        <ApproveTransactionHeader origin="0x" spenderAddress="0x" url="" />
      </Provider>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
