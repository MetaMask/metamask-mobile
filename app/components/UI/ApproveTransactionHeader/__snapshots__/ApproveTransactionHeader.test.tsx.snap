// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApproveTransactionHeader should render correctly 1`] = `
<ApproveTransactionHeader
  origin="0x"
  spenderAddress="0x"
  url=""
/>
`;
